### PR TITLE
Fix for parsing permitted name constraint for subject directory name

### PR DESCRIPTION
**Pull request by dgarske**
Cleanup to use switch in ConfirmNameConstraints.

### CODE REVIEW

File: wolfcrypt/src/asn.c [hunk 1]
```diff
@@ -4859,47 +4869,56 @@ static int ConfirmNameConstraints(Signer* signer, DecodedCert* cert)
...
-                    cert->subjectRawLen == base->nameSz &&
-                    XMEMCMP(cert->subjectRaw, base->name, base->nameSz) == 0) {
-
-                    matchDir = 1;
+                case ASN_DIR_TYPE:
+                {
+                    /* allow permitted dirName smaller than actual subject */
+                    needDir = 1;
+                    if (cert->subjectRaw != NULL &&
+                        cert->subjectRawLen >= base->nameSz &&
```
Comment (dgarske) [inline wolfcrypt/src/asn.c]: This was changed to allow a subject longer than the permitted directory subject name constraint.

File: wolfcrypt/src/asn.c [hunk 2]
```diff
@@ -4816,34 +4816,44 @@ static int ConfirmNameConstraints(Signer* signer, DecodedCert* cert)
...
-            }
-            else if (base->type == ASN_DIR_TYPE) {
-                if (cert->subjectRawLen == base->nameSz &&
-                    XMEMCMP(cert->subjectRaw, base->name, base->nameSz) == 0) {
-
-                    return 0;
+                case ASN_DIR_TYPE:
+                {
+                    /* allow permitted dirName smaller than actual subject */
+                    if (cert->subjectRawLen >= base->nameSz &&
```
Comment (dgarske) [inline wolfcrypt/src/asn.c]: This was changed to allow a subject longer than the permitted directory subject name constraint.

File: wolfcrypt/src/asn.c [hunk 3]
```diff
@@ -5417,27 +5437,37 @@ static int DecodeSubtree(byte* input, int sz, Base_entry** head, void* heap)
...
+        bType = (b & ASN_TYPE_MASK);
+
+        if (bType == ASN_DNS_TYPE || bType == ASN_RFC822_TYPE ||
+                                                        bType == ASN_DIR_TYPE) {
+            Base_entry* entry;
 
-            Base_entry* entry = (Base_entry*)XMALLOC(sizeof(Base_entry),
-                                                    heap, DYNAMIC_TYPE_ALTNAME);
+            /* if constructed has leading sequence */
+            if (b & ASN_CONSTRUCTED) {
```
Comment (dgarske) [inline wolfcrypt/src/asn.c]: A constructed directory subject name has leading sequence which needs to be skipped over for the name constraint memcmp to work.

File: wolfcrypt/src/asn.c [hunk 4]
```diff
@@ -5612,8 +5642,8 @@ static int DecodePolicyOID(char *out, word32 outSz, byte *in, word32 inSz)
             return ASN_PARSE_E;
         }
 
-        /* Validate total length (2 is the CERT_POLICY_OID+SEQ) */
-        if ((total_length + 2) != sz) {
+        /* Validate total length */
+        if (total_length > (sz - (int)idx)) {
```
Comment (dgarske) [inline wolfcrypt/src/asn.c]: This total_length check was broken. For my test case the total_length + 2 = 226 while sz 227, which incorrectly failed.

Comment (toddouska) [inline wolfcrypt/src/asn.c]: Are you saying the CERT_POLICY_OID + SEQ is not 2? Or that we didn't parse a byte?

File: wolfcrypt/src/asn.c [hunk 5]
```diff
@@ -5452,7 +5482,7 @@ static int DecodeSubtree(byte* input, int sz, Base_entry** head, void* heap)
 
             XMEMCPY(entry->name, &input[nameIdx], strLength);
             entry->nameSz = strLength;
-            entry->type = b & 0x0F;
+            entry->type = bType;
```
Comment (toddouska) [inline wolfcrypt/src/asn.c]: Why was this changed?

Comment (dgarske) [inline wolfcrypt/src/asn.c]: Its calculated above:

```
#define ASN_TYPE_MASK 0xF
/* Get type, LSB 4-bits */
bType = (b & ASN_TYPE_MASK);
```

File: wolfcrypt/src/asn.c (hunk 4 above)
Comment (dgarske) [inline wolfcrypt/src/asn.c]: The sz was 1 greater than (total_length + 2), so the != sz check was failing. I believe this was broken due to the ASN refactor. See 2805 second-cert-in-chain.pem for example.

Comment (toddouska) [inline wolfcrypt/src/asn.c]: This feels like a workaround for an off by one without understanding (at least for me) why there's an off by one.  Can you look into that a little please?

Comment (dgarske) [inline wolfcrypt/src/asn.c]: This section of code was originally changed in PR #336, where I resolved an issue parsing multiple items. It was later changed through a few refactors and now the comment about `2 is the CERT_POLICY_OID+SEQ` is incorrect here. The first sequence is total length. Then its a Sequence and OID in the loop. The updates in this PR are correct for handling checking max size.
